test/adam_smoke/CMakeLists.txt

```diff
@@ -1,5 +1,13 @@
 add_executable (adam_smoke_test adam_smoke_test.cpp)
 target_link_libraries(adam_smoke_test PRIVATE asl)
 
+if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+    target_compile_options (adam_smoke_test PRIVATE "/DNOMINMAX")
+    target_compile_options (adam_smoke_test PRIVATE "/permissive-")
+    target_compile_options (adam_smoke_test PRIVATE "/Zc:__cplusplus")
+    target_compile_options (adam_smoke_test PRIVATE "/W4")
+    target_compile_options (adam_smoke_test PRIVATE "/WX")
+endif()
```

Comment:
@dabrahams @camio - There must be a better way to manage compiler settings.
I want the library to require the first 3 options, and the cilents (test cases) to add the second 2 options.

Reply:
If by
> I want the library to require the first 3 options  and the cilents (test cases) to add the second 2 options.

You actually mean you want the library to be built with the first 3 options and its clients to have the 2nd 2 options imposed on them automatically due to the fact that they link the library, then you want

```cmake
target_compile_options(the_library PRIVATE  "/Zc:__cplusplus" PUBLIC "/W4" "/WX")
```

The last two become usage requirements as well as requirements for building the library. 

@camio's statement that:

> When the library writer dictates flags, the library gets tied to a snapshot of today's compilers and options making it unportable

Could be interpreted two ways: 1, you should try to use higher-level constructs than the flags for specific compilers whenever possible, or 2, you should never employ usage requirements.  I agree with the first but would disagree with the second.  I'm not sure what these flags do or whether higher-level constructs exist for them.

Reply:
I updated it to use PRIVATE/PUBLIC (more) appropriately.